lib/commands/ci.js

```diff
@@ -74,16 +75,24 @@ class CI extends ArboristWorkspaceCmd {
       )
     }
 
+    const workspacePaths = await getWorkspaces([], {
+      path: this.npm.localPrefix,
+      includeWorkspaceRoot: true,
+    })
+
     const dryRun = this.npm.config.get('dry-run')
     if (!dryRun) {
       // Only remove node_modules after we've successfully loaded the virtual
       // tree and validated the lockfile
       await time.start('npm-ci:rm', async () => {
-        const path = `${where}/node_modules`
-        // get the list of entries so we can skip the glob for performance
-        const entries = await fs.readdir(path, null).catch(() => [])
-        return Promise.all(entries.map(f => fs.rm(`${path}/${f}`,
-          { force: true, recursive: true })))
+        return await Promise.all([...workspacePaths.values()].map(async modulePath => {
+          const path = `${modulePath}/node_modules`
+          // get the list of entries so we can skip the glob for performance
+          const entries = await fs.readdir(path, null).catch(() => [])
+          return Promise.all(entries.map(f => {
+            return fs.rm(`${path}/${f}`, { force: true, recursive: true })
```

Comment:
Just curious: why is it important to empty a `node_modules` dir rather than deleting it entirely?

Reply:
I think this was done because the directory would need to be created again anyways.

Reply:
If anyone (like me) thought there was a significant difference in performance: no, there isn't.

```js
{ 'rm + mkdir': 94.04651467499998, 'readdir + rm': 97.85040412499998 }
```

```js
import fs from 'node:fs/promises'

const suite = {
  count: 40,
  delay: 100,
  async prepare() {
    const dir = `temp/t-${rand()}`
    await seed(dir)
    return dir
  },
  cases: {
    async 'rm + mkdir'(dir) {
      await fs.rm(dir, {recursive: true, force: true})
      await fs.mkdir(dir)
    },

    async 'readdir + rm'(dir) {
      const entries = await fs.readdir(dir, null).catch(() => [])
      await Promise.all(entries.map(entry => fs.rm(`${dir}/${entry}`, {recursive: true, force: true})))
    },
  },
  async run () {
    const results = {}
    const entries = Object.entries(this.cases)
    const dirs = await Promise.all(Array.from({length: this.count * entries.length}, () => this.prepare()))

    for (const [name, cb] of Object.entries(this.cases)) {
      let c = this.count
      results[name] = []
      do {
        const dir = dirs.pop()
        const t1 = process.hrtime.bigint()
        await cb(dir)
        const t2 = process.hrtime.bigint()
        results[name].push(Number(t2 - t1) / 1e6)
        await sleep(this.delay)
        c--
      } while (c)
    }

    const summary = Object.fromEntries(Object.entries(results).map(([name, results]) => [name, results.reduce((a, b) => a + b) / results.length]))
    console.log(summary)
  }
}

await suite.run()

async function seed(base = 'temp', length = 100, depth = 1) {
  await Promise.all(Array.from({length}, async () => {
    const dir = `${base}/${rand()}`
    const _length = (length / 10)|0
    await fs.mkdir(dir, {recursive: true})
    await Promise.all(Array.from({length: _length}, () => fs.writeFile(`${dir}/file-${rand()}`, rand())))

    if (depth) {
      await seed(dir, _length, depth - 1)
    }
  }))
}

function rand() {
  return Math.random().toString(36).slice(2)
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}
```

Reply:
What OS is this? Windows is almost always the reason for things like this.

Reply:
A lot of those old decisions were never benchmarked though, it is not surprising to find that they're (no longer) required.

Reply:
m3 / macos 14. I think the results will vary on other machines. nvm, I just calmed myself down with this test.

Reply:
Never apologize for bringing actual metrics to a discussion that has historically been about "vibes". It's really appreciated.